### PR TITLE
gh-111956: Add thread-safe one-time initialization.

### PR DESCRIPTION
**Pull request by colesbury**
The one-time initialization (`_PyOnceFlag`) is used in two places:

 * `Python/Python-ast.c`
 * `Python/getargs.c`

* Issue: gh-111956

### CODE REVIEW

**Comment by colesbury** [comment]
@ericsnowcurrently, when you have some time, would you please look at this?

One question: I'm unsure of what `int` value should be used here to indicate an error or success. I have gone with `0=error` and `1=success` here because that requires fewer modifications to the use sites, which already use that convention. I am unsure if `-1=error`, `0=success` would be better.

File: Python/getargs.c [hunk 1]
```diff
@@ -1925,40 +1926,27 @@ _parser_init(struct _PyArg_Parser *parser)
...
 static int
 parser_init(struct _PyArg_Parser *parser)
 {
-    // volatile as it can be modified by other threads
-    // and should not be optimized or reordered by compiler
-    if (*((volatile int *)&parser->initialized)) {
-        assert(parser->kwtuple != NULL);
-        return 1;
-    }
-    PyThread_acquire_lock(_PyRuntime.getargs.mutex, WAIT_LOCK);
```
Comment (ericsnowcurrently) [inline Python/getargs.c:1945]: This is the only place we use `_PyRuntime.getargs.mutex`, isn't it.  If so, we could remove it.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Aside from one small thing, LGTM.

**Comment by ericsnowcurrently** [comment]
> I'm unsure of what int value should be used here to indicate an error or success. I have gone with 0=error and 1=success here because that requires fewer modifications to the use sites, which already use that convention. I am unsure if -1=error, 0=success would be better.

The typical convention in the case of success vs. error is -1 for error and 0 for success.  The 0/1 convention is more for true/false situations.

**Comment by colesbury** [comment]
@ericsnowcurrently that makes sense. I'll change it to -1=error, 0=success.

**Comment by colesbury** [comment]
@ericsnowcurrently, I changed the code to use `-1=error` `0=success` and updated `asdl_c.py`/`Python-ast.c`. It ended up being a large change to that file, but fixed one bug where we ignored errors in `init_identifiers` due to a mismatch in return code assumptions.

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

FWIW, I haven't combed through the asdl_c.py changes but would expect the test suite to catch any mistakes in there.  It's also a good sign that you caught the bug in `init_identifiers()`. 😄

**Comment by ericsnowcurrently** [comment]
I'll merge this in the morning if no one beats me to it.

**Pull request merged**